massa-bootstrap/src/client_binder.rs

```diff
@@ -84,11 +87,36 @@ impl BootstrapClientBinder {
             u32::from_be_bytes_min(&meg_len_bytes, self.max_bootstrap_message_size)?.0
         };
 
-        // read message, check signature and deserialize
+        // read message, check signature and, optionally, signature of the message we sent just before and deserialize
         let message = {
             let mut sig_msg_bytes = vec![0u8; SIGNATURE_SIZE_BYTES + (msg_len as usize)];
-            sig_msg_bytes[..SIGNATURE_SIZE_BYTES]
-                .clone_from_slice(&self.prev_sig.unwrap().to_bytes());
+            if let Some(sended_message) = &self.sended_message {
+                let old_message_sig_from_server = {
+                    let mut sig_bytes = [0u8; SIGNATURE_SIZE_BYTES];
+                    self.duplex.read_exact(&mut sig_bytes).await?;
+                    Signature::from_bytes(&sig_bytes)?
+                };
+                // Check if old signature match
+                {
+                    let old_message = &sended_message.to_bytes_compact()?;
+                    let mut old_sig_msg_bytes =
+                        vec![0u8; SIGNATURE_SIZE_BYTES + (old_message.len())];
+                    old_sig_msg_bytes[..SIGNATURE_SIZE_BYTES]
+                        .clone_from_slice(&self.prev_sig.unwrap().to_bytes());
+                    old_sig_msg_bytes[SIGNATURE_SIZE_BYTES..].clone_from_slice(&old_message);
+                    let old_msg_hash = Hash::compute_from(&old_sig_msg_bytes);
+                    verify_signature(
+                        &old_msg_hash,
+                        &old_message_sig_from_server,
+                        &self.remote_pubkey,
+                    )?;
+                };
```

Comment:
I think this scope useless

Reply:
I made it to gain clarity and avoid using the variable in the rest of the function in case someone else develop something new because it can be confusing with the similar names etc.. I think it don't change anything when compiling and IMO bring clarity